The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/11402512076/job/31727672078) [(plain)](https://github.com/rust-lang/rust/commit/58131961a6cefc3fa5c6b1c07a066f602d5bf162/checks/31727672078/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 2.822 Building wheels for collected packages: reuse
#16 2.823   Building wheel for reuse (pyproject.toml): started
#16 3.067   Building wheel for reuse (pyproject.toml): finished with status 'done'
#16 3.068   Created wheel for reuse: filename=reuse-4.0.3-cp310-cp310-manylinux_2_35_x86_64.whl size=132720 sha256=026f3bb0f1aa8090b861fd0a0939cb1a782396d84c8aab7875096557d637a0f6
#16 3.069   Stored in directory: /tmp/pip-ephem-wheel-cache-ujsseklj/wheels/3d/8d/0a/e0fc6aba4494b28a967ab5eaf951c121d9c677958714e34532
#16 3.071 Installing collected packages: boolean-py, binaryornot, tomlkit, reuse, python-debian, markupsafe, license-expression, jinja2, chardet, attrs
#16 3.467 Successfully installed attrs-23.2.0 binaryornot-0.4.4 boolean-py-4.0 chardet-5.2.0 jinja2-3.1.4 license-expression-30.3.0 markupsafe-2.1.5 python-debian-0.1.49 reuse-4.0.3 tomlkit-0.13.0
#16 3.467 WARNING: Running pip as the 'root' user can result in broken permissions and conflicting behaviour with the system package manager. It is recommended to use a virtual environment instead: https://pip.pypa.io/warnings/venv
#16 DONE 3.6s
---

error: unused doc comment
  --> std/src/sys/alloc/windows.rs:85:5
   |
85 |       /// SAFETY: GetProcessHeap simply returns a valid handle or NULL so is always safe to call.
86 | /     unsafe {
87 | |         GetProcessHeap()
88 | |     }
   | |_____- rustdoc does not generate documentation for expressions
---

error[E0599]: no method named `is_null` found for unit type `()` in the current scope
  --> std/src/sys/alloc/windows.rs:98:40
   |
98 |     if core::intrinsics::unlikely(heap.is_null()) {
   |                                        ^^^^^^^ method not found in `()`
error[E0308]: mismatched types
   --> std/src/sys/alloc/windows.rs:102:24
    |
    |
102 |     unsafe { HeapAlloc(heap, flags, bytes) }
    |              --------- ^^^^ expected `*mut c_void`, found `()`
    |              arguments to this function are incorrect
    |
    = note: expected raw pointer `*mut core::ffi::c_void`
                 found unit type `()`
                 found unit type `()`
note: function defined here
   --> std/src/sys/alloc/windows.rs:40:51
    |
40  | windows_targets::link!("kernel32.dll" "system" fn HeapAlloc(hheap: c::HANDLE, dwflags: u32, dwbytes: usize) -> *mut c_void);

error[E0308]: mismatched types
   --> std/src/sys/alloc/windows.rs:199:27
    |
    |
199 |         unsafe { HeapFree(heap, 0, block.cast::<c_void>()) };
    |                  -------- ^^^^ expected `*mut c_void`, found `()`
    |                  arguments to this function are incorrect
    |
    = note: expected raw pointer `*mut core::ffi::c_void`
                 found unit type `()`
                 found unit type `()`
note: function defined here
   --> std/src/sys/alloc/windows.rs:82:51
    |
82  | windows_targets::link!("kernel32.dll" "system" fn HeapFree(hheap: c::HANDLE, dwflags: u32, lpmem: *const c_void) -> c::BOOL);

error[E0308]: mismatched types
   --> std/src/sys/alloc/windows.rs:212:34
    |
    |
212 |             unsafe { HeapReAlloc(heap, 0, ptr.cast::<c_void>(), new_size).cast::<u8>() }
    |                      ----------- ^^^^ expected `*mut c_void`, found `()`
    |                      arguments to this function are incorrect
    |
    = note: expected raw pointer `*mut core::ffi::c_void`
                 found unit type `()`
                 found unit type `()`
note: function defined here
   --> std/src/sys/alloc/windows.rs:61:51
    |
61  | windows_targets::link!("kernel32.dll" "system" fn HeapReAlloc(

Some errors have detailed explanations: E0308, E0599.
For more information about an error, try `rustc --explain E0308`.
error: could not compile `std` (lib) due to 7 previous errors
```

</details>

--- a/library/std/src/sys/alloc/windows.rs
+++ b/library/std/src/sys/alloc/windows.rs
@@ -3,7 +3,6 @@ use crate::alloc::{GlobalAlloc, Layout, System};
 use crate::ffi::c_void;
 use crate::mem::MaybeUninit;
 use crate::ptr;
-use crate::sync::atomic::{AtomicPtr, Ordering};
 use crate::sys::c;
 
 #[cfg(test)]
@@ -81,45 +80,9 @@ windows_targets::link!("kernel32.dll" "system" fn HeapReAlloc(
 // See https://docs.microsoft.com/windows/win32/api/heapapi/nf-heapapi-heapfree
 windows_targets::link!("kernel32.dll" "system" fn HeapFree(hheap: c::HANDLE, dwflags: u32, lpmem: *const c_void) -> c::BOOL);
 
-// Cached handle to the default heap of the current process.
-// Either a non-null handle returned by `GetProcessHeap`, or null when not yet initialized or `GetProcessHeap` failed.
-static HEAP: AtomicPtr<c_void> = AtomicPtr::new(ptr::null_mut());
-
-// Get a handle to the default heap of the current process, or null if the operation fails.
-// If this operation is successful, `HEAP` will be successfully initialized and contain
-// a non-null handle returned by `GetProcessHeap`.
-#[inline]
-fn init_or_get_process_heap() -> c::HANDLE {
-    // `HEAP` has not yet been successfully initialized
-    let heap = unsafe { GetProcessHeap() };
-    if !heap.is_null() {
-        // SAFETY: No locking is needed because within the same process,
-        // successful calls to `GetProcessHeap` will always return the same value, even on different threads.
-        HEAP.store(heap, Ordering::Release);
-
-        // SAFETY: `HEAP` contains a non-null handle returned by `GetProcessHeap`
-        heap
-    } else {
-        // Could not get the current process heap.
-        ptr::null_mut()
-    }
-}
-
-/// This is outlined from `process_heap_alloc` so that `process_heap_alloc`
-/// does not need any stack allocations.
-#[inline(never)]
-#[cold]
-extern "C" fn process_heap_init_and_alloc(
-    _heap: MaybeUninit<c::HANDLE>, // We pass this argument to match the ABI of `HeapAlloc`
-    flags: u32,
-    bytes: usize,
-) -> *mut c_void {
-    let heap = init_or_get_process_heap();
-    if core::intrinsics::unlikely(heap.is_null()) {
-        return ptr::null_mut();
-    }
-    // SAFETY: `heap` is a non-null handle returned by `GetProcessHeap`.
-    unsafe { HeapAlloc(heap, flags, bytes) }
+fn get_process_heap() -> *mut c_void {
+    // SAFETY: GetProcessHeap simply returns a valid handle or NULL so is always safe to call.
+    unsafe { GetProcessHeap() }
 }
 
 #[inline(never)]
@@ -128,20 +91,12 @@ fn process_heap_alloc(
     flags: u32,
     bytes: usize,
 ) -> *mut c_void {
-    let heap = HEAP.load(Ordering::Relaxed);
-    if core::intrinsics::likely(!heap.is_null()) {
-        // SAFETY: `heap` is a non-null handle returned by `GetProcessHeap`.
-        unsafe { HeapAlloc(heap, flags, bytes) }
-    } else {
-        process_heap_init_and_alloc(MaybeUninit::uninit(), flags, bytes)
+    let heap = get_process_heap();
+    if core::intrinsics::unlikely(heap.is_null()) {
+        return ptr::null_mut();
     }
-}
-
-// Get a non-null handle to the default heap of the current process.
-// SAFETY: `HEAP` must have been successfully initialized.
-#[inline]
-unsafe fn get_process_heap() -> c::HANDLE {
-    HEAP.load(Ordering::Acquire)
+    // SAFETY: `heap` is a non-null handle returned by `GetProcessHeap`.
+    unsafe { HeapAlloc(heap, flags, bytes) }
 }
 
 // Header containing a pointer to the start of an allocated block.
@@ -232,9 +187,9 @@ unsafe impl GlobalAlloc for System {
             }
         };
 
-        // SAFETY: because `ptr` has been successfully allocated with this allocator,
-        // `HEAP` must have been successfully initialized.
-        let heap = unsafe { get_process_heap() };
+        // because `ptr` has been successfully allocated with this allocator,
+        // there must be a valid process heap.
+        let heap = get_process_heap();
 
         // SAFETY: `heap` is a non-null handle returned by `GetProcessHeap`,
         // `block` is a pointer to the start of an allocated block.
@@ -244,9 +199,9 @@ unsafe impl GlobalAlloc for System {
     #[inline]
     unsafe fn realloc(&self, ptr: *mut u8, layout: Layout, new_size: usize) -> *mut u8 {
         if layout.align() <= MIN_ALIGN {
-            // SAFETY: because `ptr` has been successfully allocated with this allocator,
-            // `HEAP` must have been successfully initialized.
-            let heap = unsafe { get_process_heap() };
+            // because `ptr` has been successfully allocated with this allocator,
+            // there must be a valid process heap.
+            let heap = get_process_heap();
 
             // SAFETY: `heap` is a non-null handle returned by `GetProcessHeap`,
             // `ptr` is a pointer to the start of an allocated block.

